MedtrumKit/PumpManager/BluetoothManager.swift

```diff
@@ -0,0 +1,105 @@
+        guard manager.state == .poweredOn else {
+            completion(.failure(error: .invalidBluetoothState(state: manager.state)))
+            return
+        }
+
+        guard !manager.isScanning else {
+            completion(.failure(error: .alreadyScanning))
+            return
+        }
+
+        scanCompletion = completion
+        manager.scanForPeripherals(withServices: [])
+        
+        log.info("Started scanning")
+    }
+    
+    func connect(peripheral: CBPeripheral, _ completion: @escaping (MedtrumConnectResult) -> Void) {
+        if manager.isScanning {
+            manager.stopScan()
+            scanCompletion = nil
+        }
+        
+        log.info("Connecting to \(peripheral)")
+        
+        connectCompletion = completion
+        manager.connect(peripheral)
+    }
+}
+
+extension BluetoothManager {
+    func centralManagerDidUpdateState(_ central: CBCentralManager) {
+        log.info("\(String(describing: central.state.rawValue))")
+    }
+
+    func centralManager(_: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi _: NSNumber) {
+        guard let deviceName = peripheral.name, !deviceName.isEmpty else {
+            return
+        }
+        
+        // TODO: Need to validate if the device name is always MT
+        log.info("Found device: \(deviceName), \(advertisementData)!")
+        guard deviceName == "MT" else {
+            return
+        }
+        
+        // TODO: Validate processing advertismentData for Serial Number
+        let manufacturerData = advertisementData[CBAdvertisementDataManufacturerDataKey]
+        guard let manufacturerData = manufacturerData as? Data, manufacturerData.count >= 5 else {
+            log.warning("No ManufacturerData or too short - " + advertisementData.keys.joined(separator: ", "))
+            return
+        }
+        
+        scanCompletion?(.success(peripheral: peripheral, pumpSN: manufacturerData[0...4], deviceType: manufacturerData[4], version: manufacturerData[5]))
+    }
+
+    func centralManager(_: CBCentralManager, didConnect peripheral: CBPeripheral) {
+        log.info("Connected to pump: \(peripheral.name ?? "<NO_NAME>")!")
+        
+        guard let completion = connectCompletion, let pumpManager = pumpManager else {
+            return
+        }
+        
+        peripheralManager = PeripheralManager(peripheral, self, pumpManager, completion)
+        peripheral.discoverServices([PeripheralManager.SERVICE_UUID])
+    }
+
+    func centralManager(_: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
+        log.info("Device disconnected, name: \(peripheral.name ?? "<NO_NAME>")")
+    }
+
+    func centralManager(_: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
+        log.info("Device connect error, name: \(peripheral.name ?? "<NO_NAME>"), error: \(error!.localizedDescription)")
+    }
+}
```

Comment:
You might want to use 
func centralManager(_ central: CBCentralManager, willRestoreState dict: [String : Any]) {}

As well to handle coming from background, not sure how you handle this in the other loopkit pump drivers though

Reply:
This function is only needed when you provide a `CBCentralManagerOptionRestoreIdentifierKey` to the CBCentralManager.init. This allows CBCentral to skip scanning and can start directly with connecting while initializing.

Might be a nice optimization, but all other Loopkit pump drivers dont use it (and thus didnt implement it here)